hyper/http20/connection.py

```diff
@@ -525,7 +668,25 @@ def _recv_cb(self):
 
         This is generally called by a stream, not by the connection itself, and
         it's likely that streams will read a frame that doesn't belong to them.
+
+        :param stream_id: (optional) The stream ID of the stream reading data
+            from the connection.
+
         """
+        # Begin by reading what we can from the socket.
+        #
+        # Concurrency
+        #
+        # Ignore this read if some other thread has recently read data from
+        # from the requested stream
+        with self._read_lock:
+            log.debug('recv for stream %d with %s already present',
+                      stream_id,
+                      self.recent_recv_streams)
+            if stream_id in self.recent_recv_streams:
+                self.recent_recv_streams.discard(stream_id)
+                return
+
         # TODO: Re-evaluate this.
         self._single_read()
```

Comment:
I think we have a small locking issue here.

It is technically possible to yield the `_read_lock` from the block above and then be context-switched away before entering `_single_read()`. That defeats the defence given by `recent_recv_streams`. I think `_read_lock` needs to be held across the entirety of this method, with the early-exit being the short-cut most commonly used.

The other option is to do a doubly-checked mechanism where `recent_recv_streams` has its own lock to prevent blocking socket reads from stopping streams that already have data from processing that data, and then making sure we re-check that once this method takes `_read_lock` (in case the previous outstanding read that was blocked on actually returned some data for that stream).

Thoughts?